### ISSUE TITLE
Hash.new([]) dies with `Uncaught RangeError: Maximum call stack size exceeded`

### ISSUE THREAD

**Issue by graywolf** [opal/opal#1780]
Trying to use hash with default value `[]` dies with:

	Uncaught RangeError: Maximum call stack size exceeded
		at $Hash.alias [as $_initialize] (opal.min.js:1)
		at Opal.send (opal.min.js:1)
		at $Hash.TMP_Hash_initialize_60 (native.min.js:1)
		at Object.Opal.send (opal.min.js:1)
		at $Hash.alias [as $_initialize] (opal.min.js:1)
		at Opal.send (opal.min.js:1)
		at $Hash.TMP_Hash_initialize_60 (native.min.js:1)
		at Object.Opal.send (opal.min.js:1)
		at $Hash.alias [as $_initialize] (opal.min.js:1)
		at Opal.send (opal.min.js:1)
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1
	_initialize @ opal.min.js:1
	Opal.send @ opal.min.js:1
	initialize @ native.min.js:1
	Opal.send @ opal.min.js:1

Code to reproduce

	Hash.new([])

**Comment by iliabylich** [comment]
Can't reproduce it on 0.11 - http://opalrb.com/try/?code:h%20%3D%20Hash.new(%5B%5D)%20%0Ap%20h%5B%3Akey1%5D.object_id%0Ap%20h%5B%3Akey2%5D.object_id (and on the latest master), it seems to be an old bug

**Comment by graywolf** [comment]
hm, still happing on 0.11, will try to provide better repro steps if I can figure it out

**Comment by graywolf** [comment]
@iliabylich here is repo with reproduction project: https://github.com/graywolf/repro-for-opal-opal-1780

Two things though:
1) This repo assumes that having opal in one `.js` and my app in other one is supported (in that repo it's `sys.js` and `repro.js`. If that's not supported use case, this issue is invalid, I cannot reproduce it in one file.
2) The issue happens only when *both* `native` and `browser` are compiled in, so there is some kind of interaction between those those two projects (but both are opal's right?)

**Comment by graywolf** [comment]
@iliabylich dunno if you still watch closed issues so pinging by name :)

**Comment by iliabylich** [comment]
I can reproduce it locally

**Comment by iliabylich** [comment]
Ok, the reason is that `native` module gets loaded twice in your case because it was compiled twice in different modes. One is `Opal.modules['native'] = function() { ... }` (that's for `require`-ing it by name in runtime) and once as a toplevel file manually as a simple self-invoking function. Simply search for `Opal.alias(self, "_initialize", "initialize")` in your compiled file, there are two matches.

There's still a bug, minimal case for reproducing:
```
2.times do
  class Hash
    alias _initialize initialize
    def initialize(*args, &blk)
      _initialize(*args, &blk)
    end
  end
end

Hash.new({})
```

The error disappears after removing a loop and defining this method only once.

**Comment by iliabylich** [comment]
emm, this code gives exactly the same error for MRI 2.5
``` sh
$ cat test.rb
2.times do
  class M
    alias _initialize initialize
    def initialize
      _initialize
    end
  end
end

M.new

$ ruby -v test.rb
ruby 2.5.0p0 (2017-12-25 revision 61468) [x86_64-darwin17]
Traceback (most recent call last):
	10920: from test.rb:10:in `<main>'
	10919: from test.rb:10:in `new'
	10918: from test.rb:5:in `initialize'
	10917: from test.rb:5:in `initialize'
	10916: from test.rb:5:in `initialize'
	10915: from test.rb:5:in `initialize'
	10914: from test.rb:5:in `initialize'
	10913: from test.rb:5:in `initialize'
	 ... 10908 levels...
	    4: from test.rb:5:in `initialize'
	    3: from test.rb:5:in `initialize'
	    2: from test.rb:5:in `initialize'
	    1: from test.rb:5:in `initialize'
test.rb:5:in `initialize': stack level too deep (SystemStackError)
```

@elia Any ideas why it happens? That's how MRI works, so there's no bug, but I don't get why it works this way.

**Comment by iliabylich** [comment]
@graywolf The simplest solution seems to have a separate file called, let's say, `native-and-browser.rb` (that's up to you) that does `require 'native'; require 'browser'` and use it as "root" file for compiling your code. The `native` module has some troubles when you run it multiple times. Also most probably this issue can be easily reproduced by *loading* `native` multiple times (instead of requiring it):

``` ruby
require 'native'
load 'native'
Hash.new
# => RangeError: Maximum call stack size exceeded
```

To sum up: compiling all dependencies into separate files doesn't let Opal to require them only once it runtime. In some cases (like this one) it can be even unsafe, in most cases it's just slower

**Comment by graywolf** [comment]
So it seems that fix for my case is to just leave out the `native` completely and rely on `browser` providing it for me.

> In some cases (like this one) it can be even unsafe, in most cases it's just slower

so is there a way to not have separate file while still keeping recompilation time low?

	[wolf@devel tmp]$ cat out.rb 
	puts "test"
	[wolf@devel tmp]$ time opal -O -c out.rb > /dev/null

	real	0m0.284s
	user	0m0.252s
	sys	0m0.030s
	[wolf@devel tmp]$ cat in.rb 
	require 'opal'
	require 'native'
	require 'browser'

	puts "test"
	[wolf@devel tmp]$ time opal -O -c in.rb > /dev/null

	real	0m2.273s
	user	0m2.211s
	sys	0m0.051s

I don't really care that much for size (this is browser extension, so download time doesn't matter), I have it separate mainly for faster compilation time.

**Comment by iliabylich** [comment]
> What about the speed though, will it be slower this way too?

No, in this case `require` will automatically handle duplicates, that's exactly how it should look for you case

**Comment by graywolf** [comment]
Cool, will rework my projects that way than. Thank you very much for all your help `^_^`

**Comment by graywolf** [comment]
And the `sys.min.js` created this way is 250kb smaller, that's nice too.
